File: src/271_encode_and_decode_strings/encode_and_decode_strings.cpp
#include <cassert>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
   public:
    /**
     * Encodes a list of strings to a single string.
     *
     * @param strs List of strings to encode
     * @return Encoded string
     */
    string encode(vector<string>& strs) {
        string encoded = "";
        string delimiter = "#";

        for (string s : strs) {
            encoded = encoded + to_string(s.size()) + delimiter + s;
        }

        return encoded;
    }

    /**
     * Decodes a single string to a list of strings.
     *
     * @param s Encoded string
     * @return List of decoded strings
     */
    vector<string> decode(string s) {
        char delimiter = '#';
        vector<string> decoded;
        int pointer = 0;
        int end = 0;

        while (pointer < s.size()) {
            while (s[pointer] != delimiter) {
                pointer += 1;
            }
            int start = pointer + 1;

            int nextLength = stoi(s.substr(end, pointer - end));
            end = start + nextLength;

            decoded.push_back(s.substr(start, nextLength));
            pointer = end;
        }
        return decoded;
    }
};

void test_encode_decode_strings() {
    Solution solution;

    // Test case 1: Basic example
    vector<string> strs1 = {"Hello", "World"};
    string encoded1 = solution.encode(strs1);
    vector<string> decoded1 = solution.decode(encoded1);
    cout << "Test 1: [";
    for (size_t i = 0; i < strs1.size(); i++) {
        cout << "\"" << strs1[i] << "\"";
        if (i < strs1.size() - 1) cout << ", ";
    }
    cout << "] -> \"" << encoded1 << "\" -> [";
    for (size_t i = 0; i < decoded1.size(); i++) {
        cout << "\"" << decoded1[i] << "\"";
        if (i < decoded1.size() - 1) cout << ", ";
    }
    cout << "]" << endl;
    assert(decoded1 == strs1);

    // Test case 2: Empty string
    vector<string> strs2 = {""};
    string encoded2 = solution.encode(strs2);
    vector<string> decoded2 = solution.decode(encoded2);
    cout << "Test 2: [\"\"] -> \"" << encoded2 << "\" -> [\"" << decoded2[0] << "\"]" << endl;
    assert(decoded2 == strs2);

    // Test case 3: Empty list
    vector<string> strs3 = {};
    string encoded3 = solution.encode(strs3);
    vector<string> decoded3 = solution.decode(encoded3);
    cout << "Test 3: [] -> \"" << encoded3 << "\" -> []" << endl;
    assert(decoded3 == strs3);

    // Test case 4: Strings with special characters
    vector<string> strs4 = {"a#b", "c#d"};
    string encoded4 = solution.encode(strs4);
    vector<string> decoded4 = solution.decode(encoded4);
    cout << "Test 4: [\"a#b\", \"c#d\"] -> \"" << encoded4 << "\" -> [\"a#b\", \"c#d\"]" << endl;
    assert(decoded4 == strs4);

    // Test case 5: Multiple empty strings
    vector<string> strs5 = {"", "", ""};
    string encoded5 = solution.encode(strs5);
    vector<string> decoded5 = solution.decode(encoded5);
    cout << "Test 5: [\"\", \"\", \"\"] -> \"" << encoded5 << "\" -> [\"\", \"\", \"\"]" << endl;
    assert(decoded5 == strs5);

    // Test case 6: Strings with numbers
    vector<string> strs6 = {"123", "456", "789"};
    string encoded6 = solution.encode(strs6);
    vector<string> decoded6 = solution.decode(encoded6);
    cout << "Test 6: [\"123\", \"456\", \"789\"] -> \"" << encoded6
         << "\" -> [\"123\", \"456\", \"789\"]" << endl;
    assert(decoded6 == strs6);

    // Test case 7: Mixed length strings
    vector<string> strs7 = {"a", "ab", "abc", "abcd"};
    string encoded7 = solution.encode(strs7);
    vector<string> decoded7 = solution.decode(encoded7);
    cout << "Test 7: [\"a\", \"ab\", \"abc\", \"abcd\"] -> \"" << encoded7
         << "\" -> [\"a\", \"ab\", \"abc\", \"abcd\"]" << endl;
    assert(decoded7 == strs7);

    cout << "All test cases passed!" << endl;
}

int main() {
    test_encode_decode_strings();
    return 0;
}

File: src/271_encode_and_decode_strings/EncodeAndDecodeStrings.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class EncodeAndDecodeStrings {
    /**
     * Encodes a list of strings to a single string.
     * 
     * @param strs List of strings to encode
     * @return Encoded string
     */
    public String encode(List<String> strs) {
        String delimiter = "#";
        String encoded = "";

        for (String s : strs) {
            encoded = encoded + s.length() + delimiter + s;
        }

        return encoded;
    }

    /**
     * Decodes a single string to a list of strings.
     * 
     * @param s Encoded string
     * @return List of decoded strings
     */
    public List<String> decode(String s) {
        char delimiter = '#';
        int end = 0;
        int pointer = 0;
        List<String> decoded = new ArrayList<>();

        while (pointer < s.length()) {
            while (s.charAt(pointer) != delimiter) {
                pointer += 1;
            }
            int start = pointer + 1;
            int nextLength = Integer.parseInt(s.substring(end, pointer));
            end = start + nextLength;
            decoded.add(s.substring(start, end));

            pointer = end;
        }

        return decoded;
    }

    public static void main(String[] args) {
        EncodeAndDecodeStrings solution = new EncodeAndDecodeStrings();

        // Test case 1: Basic example
        List<String> strs1 = Arrays.asList("Hello", "World");
        String encoded1 = solution.encode(strs1);
        List<String> decoded1 = solution.decode(encoded1);
        System.out.println("Test 1: " + strs1 + " -> " + encoded1 + " -> " + decoded1);
        assert decoded1.equals(strs1) : "Expected " + strs1 + ", got " + decoded1;

        // Test case 2: Empty string
        List<String> strs2 = Arrays.asList("");
        String encoded2 = solution.encode(strs2);
        List<String> decoded2 = solution.decode(encoded2);
        System.out.println("Test 2: " + strs2 + " -> " + encoded2 + " -> " + decoded2);
        assert decoded2.equals(strs2) : "Expected " + strs2 + ", got " + decoded2;

        // Test case 3: Empty list
        List<String> strs3 = new ArrayList<>();
        String encoded3 = solution.encode(strs3);
        List<String> decoded3 = solution.decode(encoded3);
        System.out.println("Test 3: " + strs3 + " -> " + encoded3 + " -> " + decoded3);
        assert decoded3.equals(strs3) : "Expected " + strs3 + ", got " + decoded3;

        // Test case 4: Strings with special characters
        List<String> strs4 = Arrays.asList("a#b", "c#d");
        String encoded4 = solution.encode(strs4);
        List<String> decoded4 = solution.decode(encoded4);
        System.out.println("Test 4: " + strs4 + " -> " + encoded4 + " -> " + decoded4);
        assert decoded4.equals(strs4) : "Expected " + strs4 + ", got " + decoded4;

        // Test case 5: Multiple empty strings
        List<String> strs5 = Arrays.asList("", "", "");
        String encoded5 = solution.encode(strs5);
        List<String> decoded5 = solution.decode(encoded5);
        System.out.println("Test 5: " + strs5 + " -> " + encoded5 + " -> " + decoded5);
        assert decoded5.equals(strs5) : "Expected " + strs5 + ", got " + decoded5;

        // Test case 6: Strings with numbers
        List<String> strs6 = Arrays.asList("123", "456", "789");
        String encoded6 = solution.encode(strs6);
        List<String> decoded6 = solution.decode(encoded6);
        System.out.println("Test 6: " + strs6 + " -> " + encoded6 + " -> " + decoded6);
        assert decoded6.equals(strs6) : "Expected " + strs6 + ", got " + decoded6;

        // Test case 7: Mixed length strings
        List<String> strs7 = Arrays.asList("a", "ab", "abc", "abcd");
        String encoded7 = solution.encode(strs7);
        List<String> decoded7 = solution.decode(encoded7);
        System.out.println("Test 7: " + strs7 + " -> " + encoded7 + " -> " + decoded7);
        assert decoded7.equals(strs7) : "Expected " + strs7 + ", got " + decoded7;

        System.out.println("All test cases passed!");
    }
}
